### PR TITLE
Further XC_functionals added

### PR DESCRIPTION
**Pull request by Mdmorshadurrahman**

### CODE REVIEW

File: openmxparser/parser.py [hunk 1]
```diff
@@ -107,6 +108,24 @@ def parse(self, mainfile: str, archive: EntryArchive, logger):
             if mainfile_parser.get('scf_hubbard_u').lower == 'on':
                 method.electronic_structure_method = 'DFT+U'
 
+        section_XC_functionals1 = method.m_create(xc_functionals)
+        section_XC_functionals2 = method.m_create(xc_functionals)
+
+        scf_XcType = mainfile_parser.get('scf_XcType')
+        if scf_XcType is not None:
+            if mainfile_parser.get('scf_XcType').upper() == 'GGA-PBE':
```
Comment (ondracka) [inline openmxparser/parser.py:116]: There is no need to call the get method again. We already have the results in scf_XcType. Same for the cases below.

File: openmxparser/parser.py [hunk 2]
```diff
@@ -107,6 +108,24 @@ def parse(self, mainfile: str, archive: EntryArchive, logger):
...
+                section_XC_functionals1.XC_functional_name = 'GGA_C_PBE'
+                section_XC_functionals2.XC_functional_name = 'GGA_X_PBE'
+            elif mainfile_parser.get('scf_XcType').upper() == 'LDA' or 'LSDA-CA':
```
Comment (ondracka) [inline openmxparser/parser.py:119]: This is not doing what you think it does, the `or 'LSDA-CA'` part will always evaluate to True

**Comment by ondracka** [comment]
So first of all this is not based on top of the latest ondracka/master, please rebase.

Also instead of this quite big if-elif-... construct, the code could be simplified with a usage of dictionary like this:
```
xc_functional_dictionary = {
    'GGA-PBE': ['GGA_C_PBE','GGA_X_PBE'],
    'LDA': ['LDA_X', 'LDA_C_PZ']
    ......
}
```

and than do just
```
scf_XcType = mainfile_parser.get('scf_XcType')
for xc in xc_functional_dictionary[scf_XcType]:
    method.m_create(xc_functionals).XC_functional_name = xc
```

It looks also like we don't need the .upper() method as the functional names must be uppercase otherwise OpenMX complains (but this was my fault for not checking earlier), so we could in theory just add key "None: ['LDA_X', 'LDA_C_PZ']" to the dictionary as well and get rid of the if `if scf_XcType is not None:` check.

Tests are not needed right now as we don't have any testcase which exercises this. I'll try to add something.

**Review by ondracka — CHANGES_REQUESTED** [review]
Actually I'll try to commit a test case for the LDA stuff in a moment, so rebase on top of it and then also add tests.

File: openmxparser/parser.py (hunk 1 above)
Comment (ondracka) [reply openmxparser/parser.py:116]: This comment might be obsolete if we go for the dictionary-based rewrite. Same below.

**Comment by ondracka** [comment]
Hi Joseph,

it looks like you are a bit lost at the moment and the rebase went really wrong. Instead of trying to fix it in git which could be tricky, it might be the best if you:
1) save the changes (the final files openmx/parser.py and tests/test_parser.py files somewhere outside the git)
2) delete this branch altogether
3) switch to your master
4) rebase to latest ondracka/master
5) make a new branch (for example XC_functionals2) and switch to it
6) copy the relevant parts form the backed-up files
7) create a new commit
8) push to your repo
9) create a completely new pull request

Let me know if you run into issues.

**Comment by Mdmorshadurrahman** [comment]
> So first of all this is not based on top of the latest ondracka/master, please rebase.
> 
> Also instead of this quite big if-elif-... construct, the code could be simplified with a usage of dictionary like this:
> 
> ```
> xc_functional_dictionary = {
>     'GGA-PBE': ['GGA_C_PBE','GGA_X_PBE'],
>     'LDA': ['LDA_X', 'LDA_C_PZ']
>     ......
> }
> ```
> 
> and than do just
> 
> ```
> scf_XcType = mainfile_parser.get('scf_XcType')
> for xc in xc_functional_dictionary[scf_XcType]:
>     method.m_create(xc_functionals).XC_functional_name = xc
> ```
> 
> It looks also like we don't need the .upper() method as the functional names must be uppercase otherwise OpenMX complains (but this was my fault for not checking earlier), so we could in theory just add key "None: ['LDA_X', 'LDA_C_PZ']" to the dictionary as well and get rid of the if `if scf_XcType is not None:` check.
> 
> Tests are not needed right now as we don't have any testcase which exercises this. I'll try to add something.

If we do not use `if scf_XcType is not None:`  then how can we detect **key** when there is no 'scf_XcType'.
and for `method.m_create(xc_functionals).XC_functional_name = xc` , will it be able to place both values in two individial XC_functionals1 and 2?

**Comment by ondracka** [comment]
> If we do not use `if scf_XcType is not None:` then how can we detect **key** when there is no 'scf_XcType'.
> and for `method.m_create(xc_functionals).XC_functional_name = xc` , will it be able to place both values in two individial XC_functionals1 and 2?

If there is no 'scf.XcType' in the out file defined, i.e., the `scf_XcType = mainfile_parser.get('scf_XcType') ` is None. 
The next line is `for xc in xc_functional_dictionary[scf_XcType]:`. Now, `xc_functional_dictionary` is a dictionary, see:
https://www.w3schools.com/python/python_dictionaries.asp
so if we define it like this:
xc_functional_dictionary = {
    'GGA-PBE': ['GGA_C_PBE','GGA_X_PBE'],
    'LDA': ['LDA_X', 'LDA_C_PZ'],
    ..... orther LSDA functionals....
    None: ['LDA_X', 'LDA_C_PZ']
}
it returns the value for whatever key is specified. None is nothing special in this case, and we can use it as a key: https://stackoverflow.com/questions/7030029/is-it-reasonable-to-use-none-as-a-dictionary-key-in-python

The actual dictionary values are lists with two items. So when we do 
`for xc in xc_functional_dictionary[scf_XcType]:`
it is essentially a loop over the two items of the corresponding list for the specific key from the dictionary, so the following line `method.m_create(xc_functionals).XC_functional_name = xc` is called once for every values from the list and you get the desired behaviour.

**Comment by Mdmorshadurrahman** [comment]
> > If we do not use `if scf_XcType is not None:` then how can we detect **key** when there is no 'scf_XcType'.
> > and for `method.m_create(xc_functionals).XC_functional_name = xc` , will it be able to place both values in two individial XC_functionals1 and 2?
> 
> If there is no 'scf.XcType' in the out file defined, i.e., the `scf_XcType = mainfile_parser.get('scf_XcType') ` is None.
> The next line is `for xc in xc_functional_dictionary[scf_XcType]:`. Now, `xc_functional_dictionary` is a dictionary, see:
> https://www.w3schools.com/python/python_dictionaries.asp
> so if we define it like this:
> xc_functional_dictionary = {
> 'GGA-PBE': ['GGA_C_PBE','GGA_X_PBE'],
> 'LDA': ['LDA_X', 'LDA_C_PZ'],
> ..... orther LSDA functionals....
> None: ['LDA_X', 'LDA_C_PZ']
> }
> it returns the value for whatever key is specified. None is nothing special in this case, and we can use it as a key: https://stackoverflow.com/questions/7030029/is-it-reasonable-to-use-none-as-a-dictionary-key-in-python
> 
> The actual dictionary values are lists with two items. So when we do
> `for xc in xc_functional_dictionary[scf_XcType]:`
> it is essentially a loop over the two items of the corresponding list for the specific key from the dictionary, so the following line `method.m_create(xc_functionals).XC_functional_name = xc` is called once for every values from the list and you get the desired behaviour.

I understood the behaviour of `method.m_create(xc_functionals).XC_functional_name = xc`  and i am doing so but my only silly query is as we have two XC_functionals like:
        section_XC_functionals1 = method.m_create(xc_functionals)
        section_XC_functionals2 = method.m_create(xc_functionals) 
so, will both functional1 and functional2 will get desired values accordingly here?

**Comment by ondracka** [comment]
> I understood the behaviour of `method.m_create(xc_functionals).XC_functional_name = xc` and i am doing so but my only silly query is as we have two XC_functionals like:
> section_XC_functionals1 = method.m_create(xc_functionals)
> section_XC_functionals2 = method.m_create(xc_functionals)
> so, will both functional1 and functional2 will get desired values accordingly here?

Yes, it will work, section_XC_functionals1 and section_XC_functionals2 are just some arbitrary temporary variables, so it doesn't matter if we assign to method.m_create(xc_functionals).XC_functional_name directly.

**Pull request closed without merging**